vm/src/memory/gc.cpp

```diff
@@ -0,0 +1,187 @@
+            debug_file << "boxed int64 " << data;
+        } else if constexpr (std::is_same_v<T, double>) {
+            debug_file << "boxed double " << data;
+        } else if constexpr (std::is_same_v<T, obj_vector>) {
+            obj_vector &vec = static_cast<obj_vector &>(data);
+            debug_file << "vector len:" << vec.size();
+        } else {
+            // unreachable
+            throw MemoryError("unknown object type");
+        }
+    },
+               obj.get_data());
+
+    debug_file << std::endl;
+}
+
+void gc_debug(std::string msg) {
+    if (!debug_file.is_open())
+        return;
+    debug_file << "GC: " << msg << std::endl;
+}
+
+void msh::disable_gc_debug() {
+    debug_file.close();
+}
+
+#endif
+
+void gc::run() {
+
+    cycle++;
+
+#ifndef NDEBUG
+    gc_debug("-----------");
+    gc_debug("Running cycle " + std::to_string(cycle));
+    int t0 = time(nullptr);
+#endif
+
+    std::vector<msh::obj *> roots;
+    roots.reserve(last_roots_size);
+
+    scan_constants(roots);
+    scan_exported_vars(roots);
+    scan_thread(roots);
+
+#ifndef NDEBUG
+    gc_debug(std::to_string(roots.size()) + " roots found (last cycle: " + std::to_string(last_roots_size) + ")");
+#endif
+
+    last_roots_size = roots.size();
+
+    walk_objects(std::move(roots));
+
+    size_t removed_elements = heap_space.objects.remove_if([&](msh::obj &obj) {
+        bool detached = obj.gc_cycle != cycle;
+#ifndef NDEBUG
+        if (detached)
+            debug_obj_freed(obj);
+#endif
+        return detached;
+    });
+
+    heap_space.len -= removed_elements;
+
+#ifndef NDEBUG
+    int t1 = time(nullptr);
+    gc_debug("Removed " + std::to_string(removed_elements) + " objects.");
+    gc_debug("Cycle ended in " + std::to_string(t1 - t0) + "ms");
+#endif
+}
+
+void gc::walk_objects(std::vector<msh::obj *> to_visit) {
+    while (!to_visit.empty()) {
+        msh::obj *obj = to_visit.back();
+        to_visit.pop_back();
+        if (obj->gc_cycle == cycle)
+            continue;
+
+        obj->gc_cycle = cycle;
+        std::visit([&](auto &&obj) {
+            using T = std::decay_t<decltype(obj)>;
+            if constexpr (std::is_same_v<T, msh::obj_vector>) {
+                for (msh::obj *item : (msh::obj_vector &)obj) {
+                    to_visit.push_back(item);
+                }
+            }
+        },
+                   obj->data);
+    }
+}
+
+void gc::scan_thread(std::vector<msh::obj *> &roots) {
+    std::vector<bool> &operands_refs_offsets = thread_call_stack.operands_refs_offsets;
+
+    // the offsets will be scanned backward (following frames stack iteration)
+    size_t operands_bytes_index = operands_refs_offsets.size();
+
+    for (stack_frame frame : thread_call_stack) {
+        const Locals locals = frame.locals;
+        // add locals roots
+        for (size_t obj_ref_offset : frame.function.obj_ref_offsets) {
+            msh::obj *obj_ref = (msh::obj *)locals.get_q_word(obj_ref_offset);
```

Comment:
Don't use a C-style cast here, define a proper method to get a pointer to restricted types in the locals.